reindex/throttle.go

```diff
@@ -0,0 +1,162 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Genome Research Ltd.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ ******************************************************************************/
+
+package reindex
+
+import (
+	"sync"
+	"time"
+)
+
+const millisecondsInHour = 60 * 60 * 1000
+
+// Throttler lets you run a function limited to certain times.
+type Throttler struct {
+	op             func()
+	period         time.Duration
+	requestedStop  chan struct{}
+	requriesSignal bool
+	started        bool
+	blocked        bool
+	stopped        bool
+	runner         *opRunner
+	sync.Mutex
+}
+
+// NewThrottle returns a Throttler. Setting requiresSignal to true means that
+// you must call Signal() for op to be run in the next period.
+func NewThrottle(op func(), period time.Duration, requiresSignal bool) *Throttler {
+	return &Throttler{
+		op:             op,
+		period:         period,
+		requriesSignal: requiresSignal,
+		blocked:        requiresSignal,
+	}
+}
+
+// Start starts running our op every one of our periods, but not if the
+// function is still running, and optionally only if a signal has been
+// received.
+func (t *Throttler) Start() {
+	t.Lock()
+	defer t.Unlock()
+
+	if t.started {
+		t.Unlock()
```

Comment:
The slightly neater way of doing this would be to have an unexported `stop` method, which assumes the state is locked. You'd call this method from the exported Stop method (surrounded by the appropriate locking), and here to avoid having to unlock and relock (and having something else potiially grab the lock.